packages/ai-jsx/src/react/completion.tsx

```diff
@@ -100,5 +80,35 @@ export async function* UICompletion(
     return <Component>{children.map(toComponent)}</Component>;
   }
 
-  return <AI.React>{toComponent(JSON.parse(modelResult))}</AI.React>;
+  const Element: z.Schema = z.object({
+    name: z.string(),
```

Comment:
I tried to restrict the list of `names` to `validComponents` only, but my knowledge of `zod` is failing me. @NickHeiner is it possible to create a union of literals based on a dynamic array in `zod`? I can add a `.refine`, but the model won't see it in the first pass.

Reply:
This seems like it should work, although it gives me a type error:

```ts
  const names = ['foo', 'bar'];
  const u = z.union([names.map(name => z.literal(name))]);
```

I suspect the type error is noise can can be ignored.

Reply:
I think I tried it and it was not happy that the list was dynamic. In this code we know what `names` is at compile time, but in my case we don't. AFAIK Zod also wants to support compile time checks somehow (is that what `z.infer` does?) and because of that it only supports static types/validations. I can of course be wrong.

Reply:
No that could be right. I vaguely remember trying to do this earlier and it not working. :smile: